proposals/extensions_v2.md

```diff
@@ -0,0 +1,401 @@
+        where TAccumulator : struct, INumber<TAccumulator>
+    {
+        // Original body
+    }
+}
+
+extension Enumerable
+{
+    public TResult Sum<TSource, TResult, TAccumulator>(Func<TSource, TResult> selector)
+        for IEnumerable<TSource> source
+        where TResult : struct, INumber<TResult>
+        where TAccumulator : struct, INumber<TAccumulator>
+    {
+        // Exactly the same code as original body.
+    }
+}
+```
+
+This form supports *non* extension static methods.  For example `Enumerable.Range` would migrate like so:
+
+```c#
+static class Enumerable
+{
+    public static IEnumerable<int> Range(int start, int count) { ... }
+}
+
+
+extension Enumerable
+{
+    // Exact same signature.  No 'for-clause'.
+    public static IEnumerable<int> Range(int start, int count) { ... }
+}
+```
+
+Perfect source and binary compatibility are goals here.  That means any other features that work with static classes and extensions are expected to migrate over to extensions without change.  This includes, but it not limited to other features like 'attributes'.  For example, all attributes that might be present on a `static class` should migrate unchanged to an `extension`.  Other features not mentioned here should not be inferred to not be part of this design.  By default all features should continue being compatible, and only explicitly specified deviations should be allowed.
+
+## Disambiguation
+
+Classic extension methods today can be disambiguated by falling back to static-invocation syntax.  For example, if `x.Count()` is ambiguous, it is possible to switch to some form of `StaticClass.Count(x)` to call the desired method.  A similar facility is needed for modern extension members.  While the existing method-invocation-translation approach works fine for methods (where the receiver can be remapped to the first argument of the static extension method call), it is ungainly for these other extension forms.
+
+As an initial strawman this proposal suggests reusing `cast expression` syntax for disambiguation purposes.  For example:
+
+```c#
+var v1 = ((Extension)receiver).ExtensionMethod(); // instead of Extension.ExtensionMethod(receiver)
+var v2 = ((Extension)receiver).ExtensionProperty;
+var v3 = ((Extension)receiver)[indexerArg];
+var v4 = (Extension)receiver1 + receiver2;
+```
+
+Constructors and static methods would not need any special syntax as the extension can cleanly be referenced as a type where needed.
+
+```c#
+var v3 = new Extension(...); // Makes instance of the actual extended type.
+var v4 = Extension.StaticExtensionMethod(...);
+```
+
+Note 1: while the cast syntax traditionally casts or converts a value, that would not be the case for its use here.  It would only be used as a lookup mechanism to indicate which extension gets priority.  Importantly, even with this syntax, extensions themselves are not types.  For example:
+
+```c#
+Extension e1;                   // Not legal.  Extension is not a type.
```

Comment:
>Extension is not a type. [](http://example.com/codeflow?start=47&length=24)

nit: "Extension is not a type." That seems too strong. For static types, we just say "static types cannot be used as parameters/type arguments/..."